### PR TITLE
 Authorization via OPA

### PR DESCRIPTION
**Pull request by jastBytes**
This implements issue #66.

### CODE REVIEW

File: build/package/helm/gateway/files/policies/policies_test.rego [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package m8.authz
+
+alice_admin = {"User": {"Id": "1234", "Name": "alice", "Roles": [{"Name": "admin", "Scope": "system"}, {"Name": "admin", "Scope": "tenant", "Resource": "1234"}]}}
+
+bob_tenant_admin = {"User": {"Id": "12345", "Name": "bob", "Roles": [{"Name": "admin", "Scope": "tenant"}]}}
```
[the remark this answers is not in the document]
Comment (HaniAlshikh) [reply build/package/helm/gateway/files/policies/policies_test.rego:5]: as far as i understood even in the old setup the resource was not checked and I think it is important to do so, or do we say a tenant admin is admin of all tenants? at least for now?

**Review by HaniAlshikh — CHANGES_REQUESTED** [review]

File: internal/gateway/auth_server_test.go [hunk 2]
```diff
@@ -15,108 +15,144 @@
...
-		req.Header.Set(HeaderAuthorization, fmt.Sprintf("bearer %s", signedToken))
-		res, err := env.HttpClient.Do(req)
+		conn, err := CreateInsecureConnection(ctx, env.ApiListenerAPIServer.Addr().String())
 		Expect(err).ToNot(HaveOccurred())
-		Expect(res.StatusCode).To(Equal(http.StatusUnauthorized))
+		defer conn.Close()
+		authClient := gateway.NewGatewayAuthClient(conn)
+
+		resp, err := authClient.Check(ctx, &gateway.CheckRequest{
+			FullMethodName: "/something/",
```
[the remark this answers is not in the document]
Comment (jastBytes) [reply internal/gateway/auth_server_test.go:125]: Agreed.

File: internal/gateway/auth_server_test.go [hunk 3]
```diff
@@ -15,108 +15,144 @@
...
+		Expect(err).ToNot(HaveOccurred())
+		defer conn.Close()
+		authClient := gateway.NewGatewayAuthClient(conn)
 
-		req.Header.Set(HeaderAuthorization, fmt.Sprintf("bearer %s", "notavalidjwt"))
-		res, err := env.HttpClient.Do(req)
-		Expect(err).NotTo(HaveOccurred())
-		Expect(res.StatusCode).To(Equal(http.StatusUnauthorized))
+		resp, err := authClient.Check(ctx, &gateway.CheckRequest{
+			FullMethodName: "/something/",
```
[the remark this answers is not in the document]
Comment (jastBytes) [reply internal/gateway/auth_server_test.go:99]: Agreed

File: Makefile [hunk 4]
```diff
@@ -28,18 +28,16 @@ help: ## Display this help.
 	@awk 'BEGIN {FS = ":.*##"; printf "\nUsage:\n  make \033[36m<target>\033[0m\n"} /^[a-zA-Z_0-9-]+:.*?##/ { printf "  \033[36m%-15s\033[0m %s\n", $$1, $$2 } /^##@/ { printf "\n\033[1m%s\033[0m\n", substr($$0, 5) } ' $(MAKEFILE_LIST)
 
 add-license: ## Adds the license to every file
-	@docker run --rm -v "$(PWD):/src" -u $(shell id -u) ghcr.io/google/addlicense -c "Monoskope Authors" -l "apache" -v .
+	@docker run --rm -v "$(PWD):/src" -u $(shell id -u) ghcr.io/google/addlicense --ignore "**/*.yaml" --ignore "**/*.yml" -c "Monoskope Authors" -l "apache" -v .
```
[the remark this answers is not in the document]
Comment (jastBytes) [reply Makefile:31]: Done

File: Makefile [hunk 5]
```diff
@@ -28,18 +28,16 @@ help: ## Display this help.
...
 
-check-license: ## Checks thath the license is set on every file
-	@docker run --rm -v "$(PWD):/src" -u $(shell id -u) ghcr.io/google/addlicense -c "Monoskope Authors" -l "apache" -v -check .
+check-license: ## Checks that the license is set on every file
+	@docker run --rm -v "$(PWD):/src" -u $(shell id -u) ghcr.io/google/addlicense --ignore "**/*.yaml" --ignore "**/*.yml" -c "Monoskope Authors" -l "apache" -v -check .
```
[the remark this answers is not in the document]
Comment (jastBytes) [reply Makefile:34]: Done

File: docs/diagrams/~$00-monoskope-architecture.drawio.svg.bkp [hunk 6]
```diff
@@ -0,0 +1,4 @@
...
...ing-top: 856px; margin-left: 436px;"><div data-drawio-colors="color: rgb(0, 0, 0); " style="box-sizing: border-box; font-size: 0px; text-align: center;"><div style="display: inline-block; font-size: 12px; font-family: Helvetica; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: none; font-weight: bold; white-space: nowrap;"><div>K8s</div><div>Reactor</div></div></div></div></foreignObject><text x="436" y="860" fill="rgb(0, 0, 0)" font-family="Helvetica" font-size="12px" text-anchor="middle" font-weight="bold">K8s...</text></switch></g><path d="M 363 896 L 363 906 Q 363 916 363 914.5 L 363 913.75 Q 363 913 363 919.82 L 363 926.63" fill="none" stroke="rgb(0, 0, 0)" stroke-miterlimit="10" pointer-events="none"/><path d="M 363 931.88 L 359.5 924.88 L 363 926.63 L 366.5 924.88 Z" fill="rgb(0, 0, 0)" stroke="rgb(0, 0, 0)" stroke-miterlimit="10" pointer-events="none"/></g><switch><g requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility"/><a transform="translate(0,-5)" xlink:href="https://www.diagrams.net/doc/faq/svg-export-text-problems" target="_blank"><text text-anchor="middle" font-size="10px" x="50%" y="100%">Text is not SVG - cannot display</text></a></switch></svg>
```
[the remark this answers is not in the document]
Comment (jastBytes) [reply docs/diagrams/~$00-monoskope-architecture.drawio.svg.bkp]: Done

File: internal/gateway/gateway_server.go [hunk 7]
```diff
@@ -48,87 +48,41 @@ func NewGatewayAPIServer(authConfig *auth.ClientConfig, client *auth.Client, ser
...
+	s.log.V(logger.DebugLevel).Info("User exists!", "name", user.Name, "email", user.Email, "id", user.ID())
 
 	// Override upstream name
 	upstreamClaims.Name = user.Name
 
 	// Issue token
 	signedToken, rawToken, err := s.authServer.IssueToken(ctx, upstreamClaims, user.ID().String())
 	if err != nil {
-		s.log.Error(err, "Issuing token failed.")
+		s.log.Error(err, "Issuing token failed.", user.Name, "email", user.Email, "id", user.ID())
```
[the remark this answers is not in the document]
Comment (jastBytes) [reply internal/gateway/gateway_server.go:85]: done

File: internal/gateway/oidc_provider_server.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+	apiLis, err := net.Listen("tcp", apiAddr)
+	if err != nil {
+		return err
+	}
+	defer apiLis.Close()
+
+	return s.ServeFromListener(apiLis)
+}
+
+// Serve tells the server to start listening on given listener.
```
[the remark this answers is not in the document]
Comment (jastBytes) [reply internal/gateway/oidc_provider_server.go:71]: done

File: build/package/helm/gateway/files/policies/policies_test.rego (hunk 1 above)
[the remark this answers is not in the document]
Comment (jastBytes) [reply build/package/helm/gateway/files/policies/policies_test.rego]: extended the whole thing

File: build/package/helm/gateway/files/policies/policies_test.rego [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+test_system_admin {
+	is_system_admin with input as alice_admin
+	not is_system_admin with input as bob_tenant_admin
+}
+
+test_authorized {
+	authorized with input as alice_admin
+	authorized with input as jane
+	authorized with input as scim_scope
+}
```
Comment (HaniAlshikh) [inline build/package/helm/gateway/files/policies/policies_test.rego:20]: I think we should have a test case for `tenant_admin_rolebindings` that makes sure bob can only execute on 1234 and not any other resource.

**Review by HaniAlshikh — CHANGES_REQUESTED** [review]
My Brain hearts 🙃

File: api/gateway/messages.proto [hunk 10]
```diff
@@ -42,9 +43,9 @@ message UpstreamAuthenticationResponse {
 
 message AuthenticationRequest {
   // code is the auth code received by the IDP
-  string code = 1;
+  string code = 1 [ (validate.rules).string = {min_len : 1} ];
   // state is the encoded, nonced AuthState
```
[the remark this answers is not in the document]
Comment (jastBytes) [reply api/gateway/messages.proto:47]: https://de.wikipedia.org/wiki/Nonce I mean nonced but yes it is not a real word. :D

File: internal/gateway/auth_server_test.go [hunk 11]
```diff
@@ -15,147 +15,206 @@
...
 )
 
 var _ = Describe("Gateway Auth Server", func() {
-	It("can retrieve openid conf", func() {
-		res, err := env.HttpClient.Get(fmt.Sprintf("http://%s/.well-known/openid-configuration", localAddrAuthServer))
+	var (
+		ctx              = context.Background()
+		expectedUserId   = uuid.New()
+		expectedRole     = roles.User.String()
+		exptectedScope   = scopes.Tenant.String()
```
[the remark this answers is not in the document]
Comment (jastBytes) [reply internal/gateway/auth_server_test.go:46]: Fixed

File: build/package/helm/gateway/files/policies/policies_test.rego (hunk 9 above)
Comment (jastBytes) [reply build/package/helm/gateway/files/policies/policies_test.rego:37]: Done

File: build/package/helm/gateway/files/policies/policies.rego [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+# check if system admin
+is_system_admin {
+	some role in input.User.Roles
+	role.Scope == scope_system
+	role.Name == role_admin
+	print(input.User.Name, "is system admin")
+}
+
+# check if user is tenant admin and adjusts rolebindings of other users of the tenant
```
Comment (HaniAlshikh) [inline build/package/helm/gateway/files/policies/policies.rego:33]: Can you please explain what you mean by "adjusts rolebindings ..."? As far as I understood we are only checking here, right?

**Review by HaniAlshikh — APPROVED** [review]

File: api/gateway/messages.proto (hunk 10 above)
[the remark this answers is not in the document]
Comment (HaniAlshikh) [reply api/gateway/messages.proto:47]: Oh never heard of it before 😅

File: build/package/helm/gateway/files/policies/policies.rego (hunk 12 above)
Comment (jastBytes) [reply build/package/helm/gateway/files/policies/policies.rego:33]: The comment means that this checks if the user is tenant admin and the user is trying to "change" rolebindings of other users for the tenant he or she is admin for.

**Review by HaniAlshikh — APPROVED** [review]

**Pull request merged**